### ISSUE TITLE
False Positive Detection of "no-redeclare"

### ISSUE THREAD

**Issue by FrederikNS** [eslint/eslint#4646]
I recently ran into a case where eslint detects a variable redeclaration, when a variable is declared in two mutually exclusive if-else branches. The code is a bit of a code smell, as variable hoisting could introduce issues, but I think the "no-redeclare" flag is a bit misleading, as the variable would never be declared twice.

This snippet shows the issue, and it can be run in the eslint demo here: http://eslint.org/demo/

Code:

```javascript
function noRedeclareFalsePositive() {
    var x = false;
    
    if(x) {
        var y = "X was very true";
        return y;
    } else {
        var y = "X was not so true";
        return y;
    }
}

noRedeclareFalsePositive();
```

ESLint output:

```
8:13 - "y" is already defined (no-redeclare)
```

The version is 1.10.3

**Comment by eslintbot** [comment]
@FrederikNS Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by platinumazure** [comment]
That's the thing, though: due to hoisting, it is in fact declared twice. It's only initialized once, but two declarations via hoisting is exactly what this rule is supposed to catch.

What about if block bindings (I.e., `let`) are used? Will that trigger a warning? If not, then I would say this is working as intended.

**Comment by FrederikNS** [comment]
I might be wrong here, but isn't the variable only declared when it's first used, or when explicitly declared? or does the javascript engine declare all variables in a function up front?

**Comment by nzakas** [comment]
All `var` declarations are hoisted to the top of the function. It doesn't matter if execution flows into those branches or not, the variable is always declared.

**Comment by mysticatea** [comment]
FYI: A rule to emulate block scoping for `var` exists in my plugin: https://www.npmjs.com/package/eslint-plugin-mysticatea
